website/src/app/blog/sans-io/readme.mdx

```diff
@@ -0,0 +1,558 @@
+
+sans-IO code needs to be "driven" by an eventloop of some sorts because it "just" expresses the state of the system but doesn’t cause any side-effects itself.
+The eventloop is responsible for "querying" the state (like `poll_transmit`), executing it and also passing new input to the state machine (`handle_timeout` and `handle_input`).
+To some people, this may appear as unnecessary boilerplate but it comes with a great benefit: flexibility.
+
+* Want to make use of `sendmmsg` to reduce the number of syscalls when sending packets? No problem.
+* Want to multiplex multiple protocols over a single socket? No problem.
+
+Writing the eventloop yourself is an opportunity to be able to tune our code to exactly what we want it to do.
+This also makes maintenance easier for library authors: They can focus on correctly implementing protocol functionality instead of having debates around async runtimes or exposing APIs to set socket options.
+
+A good example here is `str0m`’s stance on the gathering of candidates:
+This is an IO concern and up to the application on how to achieve it.
+`str0m` only provides an API to add such an ICE candidate to the current state.
+As a result, we are able to easily implementation optimisations such as gathering TURN candidates prior to any connection being made, thus reducing firezone's connection-setup latency.
+
+### Testing at the speed of light
+
+sans-IO code is essentially side-effect free and thus lends itself extremely well for (unit) tests.
+Due to sockets and time being abstracted away, it becomes a breeze to write tests that advance time by 5 minutes in an instant.
+All we need to do is pass a modified `Instant` to our function and assert, how the code behaves.
+To see a real world example of this, [check out](https://github.com/firezone/firezone/blob/53557f46e452c0fe5195a4326873753a356c6005/rust/connlib/snownet/tests/lib.rs#L123-L127) how we test that `snownet` closes idle connections after 5 minutes.
+
+Similarly, actually sending data over a socket takes (a little bit of) time and more importantly, requires allocation of ports etc.
+In a sans-IO world, "sending data" in a test is as simple as taking a `Transmit` from party B and calling `handle_input` on the state of party A.
+No need to go through a network socket!
+
+At firezone, we took this idea one step further.
+We implemented a reference state machine that describes how we want `connlib` to work.
+This reference state machine is used as the source of truth in our tests.
+We then leverage `proptest`'s support for [state machine testing](https://proptest-rs.github.io/proptest/proptest/state-machine.html) to deterministically sample and execute thousands of scenarios on every CI run and compare the reference state machine with `connlib`'s actual state.
+The details of this go beyond the scope of this post but we are planning to publish one about that topic in particular too!
+The key take-away here is that a sans-IO design enables these kind of tests.
```

Comment:
It might be cool to conclude this section with a screenshot of the proptest wall time in CI (a few seconds?) on the GH free runners.

Reply:
It is actually closer to a minute with level 1 optimisations. But we also run 1000 tests with 1-20 transitions each :)